### ISSUE TITLE
How to upgrade indirect dependencies?

### ISSUE THREAD

**Issue by chinesedfan** [yarnpkg/yarn#4986]
**Do you want to request a *feature* or report a *bug*?**

Feature.

**What is the current behavior?**
`yarn upgrade` ignores indirect dependencies, so users can't upgrade them in yarn.lock. If I missed something, please tell me.

**If the current behavior is a bug, please provide the steps to reproduce.**
- Suppose a new empty project, run `yarn add is-alphanumerical@1.0.0`
  - 2 indirect dependencies, `is-alphabetical` and `is-decimal`, will be installed and saved in yarn.lock
  - the latest version of `is-alphabetical` is 1.0.1 now, if another new version, say 1.0.2 was released(to test, you can release 2 test packages by yourself or modify `is-alphabetical` to be 1.0.0 in `yarn.lock`, ** I know modifying yarn.lock directly is not a regular operation**)
- No matter which of following ways, `yarn` always says `All of your dependencies are up to date`
  - yarn upgrade is-alphabetical
  - yarn upgrade-interactive
  - yarn upgrade-interactive is-alphabetical

**What is the expected behavior?**
`yarn upgrade` also supports indirect dependencies.

**Please mention your node.js, yarn and operating system version.**
Node 8.9.0
yarn 1.3.2
OSX 10.12.6

**Comment by milesj** [comment]
@chinesedfan Have you tried `yarn upgrade-interactive`?

**Comment by chinesedfan** [comment]
@milesj Yes, it results in the same result and I also updated reproduce steps in issue description.

**Comment by rally25rs** [comment]
That is because `yarn add is-alphanumerical@1.0.0` sets your package.json to _exactly_ version `1.0.0` as you requested.

`yarn upgrade` respects your package.json semver range, and since you specified exactly version 1.0.0, it won't offer to upgrade to other versions.

You could resolve this a couple ways:

* `yarn upgrade --latest` will ignore semver range and see what is tagged as `latest` in the registry.
* change package.json to accept a version range like `^1.0.0` then `yarn upgrade` (you might have to `yarn install` first to get it to update the lock file for the changed range)
* explicitly specify a version to `upgrade`, like `yarn upgrade is-alphanumerical@1.0.1` or `yarn upgrade is-alphanumerical@^1.0.0`

**Comment by chinesedfan** [comment]
@rally25rs Thanks for your reply! I tested 2 more cases.

- `yarn upgrade is-alphabetical  --force` doesn't work, either.
- `yarn upgrade is-alphanumerical` will upgrade ALL its subdependencies even if it is already latest.
  - But if I just want to upgrade a specified subdependency, that is still not very convenient.

**Comment by OneCyrus** [comment]
yes, this is a major problem with yarn at the moment. and it's already in discussion at #2394

**Comment by rally25rs** [comment]
duplicate #2394

**Comment by AlexWayfer** [comment]
Please, re-open: it's not duplicate.

#2394 describes duplicating of `meck-test-bb` package (indirect dependency):

> I got two copies of meck-test-bb

This issue describes just ability to upgrade indirect dependency (somehow).

**Comment by chinesedfan** [comment]
@rally25rs Forgive me to ping.

**Comment by AlexWayfer** [comment]
@rally25rs, please, you've closed both non-duplicating issues, it's wrong. Give us ability to upgrade indirect dependencies, please!

**Comment by rally25rs** [comment]
Sorry, there was some confusion over on the other issue. I originally thought 2394 was asking for a way to upgrade a transitive dep using a `--deep` flag, or something like that, so I had marked this issue as a duplicate of that. Later on after re-reading 2394 I think it was about something different, which is not a duplicate of this like I originally thought.

**Comment by alex-thewsey-ibm** [comment]
+1 for this feature request. Also an example for anybody dumb like me who needs to upgrade a specific indirect dependency manually in the interim:

Given explicit dependency `jsonwebtoken` has resolved implicit dependency `jws^3.0.0` to vulnerable`jws=3.1.4`: and you need it to instead resolve to patched `3.1.5`.

Delete the `jws` entry e.g. below from yarn.lock, and re-run `yarn`. The indirect dependency and any affected packages will be updated, without touching other things (on yarn v1.3 at least)

```
jws@^3.0.0, jws@^3.1.4:
  version "3.1.4"
  resolved "https://registry.npmjs.org/jws/-/jws-3.1.4.tgz#f9e8b9338e8a847277d6444b1464f61880e050a2"
  dependencies:
    base64url "^2.0.0"
    jwa "^1.1.4"
    safe-buffer "^5.0.1"
```

**Comment by mkutny** [comment]
@alex-thewsey-ibm, thanks for the workaround!

Worked on yarn v1.7.

**Comment by Subtletree** [comment]
ty, worked Yarn 1.9.2

**Comment by joelpurra** [comment]
Might help to nudge yarn with selective dependency `resolutions`, even if it's for a single dependency. Thanks to @remolueoend for the hint!
https://yarnpkg.com/lang/en/docs/selective-version-resolutions/

From the docs:

```json
{
  "name": "project",
  "version": "1.0.0",
  "dependencies": {
    "left-pad": "1.0.0",
    "c": "file:../c-1",
    "d2": "file:../d2-1"
  },
  "resolutions": {
    "d2/left-pad": "1.1.1",
    "c/**/left-pad": "1.1.2"
  }
}
```

**Comment by ai** [comment]
We need this feature in Autoprefixer to suggest users how to update `caniuse-lite` in their `yarn.lock` https://github.com/postcss/autoprefixer/issues/1184

**Comment by richseviora** [comment]
Same problem here. I expected `yarn upgrade caniuse-lite browserslist` to upgrade the sub-dependency. It didn't do that, nor did it give me an error message saying that it can't upgrade it b/c it's not a dependency.

Deleting the relevant lockfile entries and then re-running `yarn` as @alex-thewsey-ibm suggested fixed the immediate issue for me.

**Comment by jrochkind** [comment]
It is odd to me that yarn is missing tnis feature. I am new to yarn (and npm), so assumed there must be a way to do this. I'm still not completely sure if there *is* a non-obvious way to do this that none in this thread know, or if there's really no way to do this. 

If there's really no way to upgrade a transitive/indirect dependency in the lockfile without adding it to package.json... I don't understand how yarn users do without it.

**Comment by Machiaweliczny** [comment]
This bug is weird - I tried to update lodash recently with `yarn upgrade lodash@4.17.15` and it still haven't updated for any transient dependency. 

I was still left with all `major`(^4.X.X) and `patch`(~4.17.X) versions pointing to old version. 

Only way to fix it is via manual editting of `yarn.lock` and then maybe running `yarn upgrade` to consolidate changes. I would expect a little better from such widely used tool.

Is this acknowledged bug or yarn is conservative by default and I'm expected to manually edit `yarn.lock` or use some flag?

**Comment by cheshrkat** [comment]
I suspect I have the same security alert to resolve as @Machiaweliczny ;) It would be really useful to override the indirect dependency while we wait on projects to fix their own. Even with highly responsive projects there's a delay waiting for fixes and releases.

**Comment by djmitche** [comment]
Indeed, this is problematic for security issues in indirect dependencies, and the workaround of editing `yarn.lock`, while effective, is disappointing and difficult to automate.  If this is not the default behavior for some reason, it'd be great to add an option like `--include-indirect` that will force Yarn to follow the indirect dependencies.

**Comment by jrochkind** [comment]
I don't think it should need an option, I don't see why `yarn upgrade [an indirect dependency]` doesn't just update the yarn.lock, without any need for additional options. 

However, another workaround I found I am happy with is adding `resolutions` to my package.json. If `lodash` is an indirect dependency, and I know that I need it to be >= 4.7.13 to avoid a security vulnerability, I can add to my package.json:

```
  "resolutions": {
    "lodash": ">= 4.17.13"
  }
```

Then just run `yarn install`, it will update the `yarn.lock` to meet that requirement, or complain if it can't because it conflicts with an indirect dependency. 

This actually works pretty well; I wonder if it's not a "workaround" but the intended solution? It took me a while to discover though.

**Comment by milesj** [comment]
Why not yarn install the transitive dep you want to update but don't commit the package.json changes, just the yarn.lock?

**Comment by kumar303** [comment]
> @djmitche Indeed, this is problematic for security issues in indirect dependencies, and the workaround of editing `yarn.lock`, while effective, is disappointing and difficult to automate. 

This is another workaround which is slightly more automatable:

```
yarn remove is-alphanumerical
yarn add is-alphanumerical
```

Using the example in the PR description, this would remove the top level dep then re-add it which will get all of its latest sub-deps, according to the ranges specified by `is-alphanumerical` (caret ranges, for example). It will then produce a new lock file.

The side effect is that it will update *all* sub-deps which is not ideal. For a security issue in sub-dep A I'd want to *only* update sub-dep A.

The workaround of adding sub-dep A as a direct dep just to fix a security issue is worse because it creates confusion (the package is not used directly) as well as a maintenance burden.

**Comment by Cellule** [comment]
> yarn remove is-alphanumerical
yarn add is-alphanumerical

This seems to be the only reliable way to actually update a dependency. I realized today that we were stuck on a `1.0.0` version of a package that had updated to `1.1.0` a year ago. The package we were using did use `^1.0.0` and all the times we "upgraded" that package it never picked up the new `1.1.0` version of its dependency.
Turns out there was a pretty bad bug that were silently failing in our product that should have been fixed a year ago without me wasting a day investigating why we were having this issue.

I can't believe editing yarn.lock manually, removing then re-adding the package or using selective resolution are the "ways" to update an indirect dependency.

IMO, `yarn upgrade` should update the indirect dependencies to the latest accepted semver version, that's why we upgrade a package. I mean, if there's was a break in the semver range it does update the indirect dependencies, so it should do it always

**Comment by djmitche** [comment]
Would it help to write some kind of external script to do this?  I suppose it would just remove all `yarn.lock` entries for the given package, and then re-run yarn?

**Comment by pftg** [comment]
There is a simple script to do this: https://gist.github.com/pftg/fa8fe4ca2bb4638fbd19324376487f42

**Comment by nnmrts** [comment]
Can one of the maintainers please change the label from `cat-feature` to `cat-bug`?

**Comment by rally25rs** [comment]
> Can one of the maintainers please change the label from cat-feature to cat-bug?

why? this isn't a bug. It is as designed. `yarn upgrade` was never intended to be used to upgrade a  transitive dependency. The originally opened "issue" is even labeled as a feature request.

Internally `upgrade` uses `yarn outdated` to determine what is out of date and what versions to upgrade to. `outdated` only checks direct dependencies.

I could be wrong, or perhaps it has changed, but I'm pretty sure that `npm upgrade` at least as of 3 years ago when `yarn upgrade` was last reworked, also doesn't provide a way to upgrade a transitive dep. (again, that might have changed since over the years, I'm not too up-to-date on npm's changes).

-----

Anyone is free to submit a PR to add this functionality. This is an open source project and it is up to the community at large to contribute. This feature request has been open for years but no one has stepped up to implement it and that is why it hasn't been "fixed".

**Comment by pftg** [comment]
@nnmrts could you check my script https://github.com/yarnpkg/yarn/issues/4986#issuecomment-562719589 will it help you for now?

**Comment by nnmrts** [comment]
@pftg Sorry, I was tired and grumpy, consider my comment as resolved. 😬

**Comment by nnmrts** [comment]
> @nnmrts could you check my script [#4986 (comment)](https://github.com/yarnpkg/yarn/issues/4986#issuecomment-562719589) will it help you for now?

That script unfortunately didn't work for me, tried it out yesterday. Maybe I did something wrong, but my whole yarn.lock file got "emptied" by the script.

**Comment by dacioromero** [comment]
I'm not sure how good of a workaround this is, but I ran this script that I wrote:

```javascript
const fs = require('fs')
const lockfile = require('@yarnpkg/lockfile')
const package = require('./package.json')

const lock = lockfile.parse(fs.readFileSync('yarn.lock', 'utf-8')).object

const allDeps = new Set()

const parseDep = ([name, version]) => {
  allDeps.add(`${name}@${version}`)
}

Object.entries(package.dependencies).forEach(parseDep)
Object.entries(package.devDependencies).forEach(parseDep)

const newLock = Object.fromEntries(Object.entries(lock).filter(([dep]) => allDeps.has(dep)))
const newLockString = lockfile.stringify(newLock)

fs.writeFileSync('yarn.lock', newLockString)
```

Then ran `yarn install` and it seems to install the latest verisons of indirect dependencies.

**Comment by ayushya** [comment]
I was able to resolve deep/indirect dependencies successfully. I thought I should share it with everyone.

https://medium.com/@ayushya/upgrading-javascript-packages-deep-dependencies-using-yarn-8b5983d5fb6b

I have tried to resolve and explain the risks of re-generating the yarn.lock and have suggested what should be done.

Let me know if this works for you guys as well. Or any suggestions to improve the process of upgrading.

**Comment by jrochkind** [comment]
@ayushya Hm, that does seem to work, genius. 

I wonder if yarn would accept a patch in which `yarn upgrade` to an indirect dependency (or some other command?) just... did that?

**Comment by natematykiewicz** [comment]
@jrochkind I would have expected a `yarn upgrade` of an indirect dependency to upgrade it even if it's not my direct dependency. Without that feature, you can be years behind on updates of indirect dependencies.

In my case, I was trying to upgrade `fsevents`, so that it didn't spew errors when I did a `yarn install`. `fsevents` isn't a package I directly use -- it's something that `webpack-dev-server` uses. But shockingly, I was locked to whatever version existed when `webpack-dev-server` was first installed in this app.

I had to use this trick to upgrade it, which seemed like a total hack. https://github.com/yarnpkg/yarn/issues/4986#issuecomment-395036563

**Comment by FelipeLujan** [comment]
the workaround is not working for me, unfortunately, the old deep-dependencies were added again to the yarn.lock file after deleting them. I can see them back in the node_modules folder and yarn.lock file after running yarn install.

maybe the workaround is not compatible with yarn workspaces?

**Comment by jrochkind** [comment]
@FelipeLujan when the workaround works, the deep-dependencies are still added again to the yarn.lock file -- that is expected -- but with new later versions. But only with new versions if there *are* new versions released, *and* if they are allowed by the dependency tree. If some intermediary dependency expresses a restriction which doesn't allow an upgrade, they still can't be upgraded. They are just upgraded to the latest allowed by the restrictions in the tree. 

i don't use yarn workspaces though, so can't say if that's relevant.

**Comment by ayushya** [comment]
@FelipeLujan AFAIK yarn workspaces work in a similar way.

The workaround to delete the package section will only upgrade the package to the latest **MINOR/PATCH** version.
If you want to upgrade the Package to a newer **MAJOR** version you have to find the package dependency chain running `yarn why package-name-here` and upgrade the package at the top of its chain.

CAUTION: Do test your code as upgrading the packages to a newer MAJOR version might bring some breaking changes.

**Comment by djmitche** [comment]
https://github.com/djmitche/yarn-minify might help with this..

**Comment by jrochkind** [comment]
@ayushya's workaround is working great for me, manually editing the yarn.lock to remove an indirect dependency, then running yarn install to re-add it at a later version. 

To me, this seems like a feature that should be built into yarn though, rather than requiring you to manually edit yarn.lock. It seems like it should be fairly striaghtforward to make a feature that is as if you had manually edited to delete the dependency and run yarn install. I feel this feature really oughta be built into yarn, and am kinda confused as to why it isn't.

**Comment by pramoditagi** [comment]
> I was able to resolve deep/indirect dependencies successfully. I wonder when we will get an official support.
> 
> https://medium.com/@ayushya/upgrading-javascript-packages-deep-dependencies-using-yarn-8b5983d5fb6b
> 
> I have tried to resolve and explained the risks of re-generating the yarn.lock and have suggested what should be done.
> 
> Let me know if this works for you guys as well. Or any suggestions to improve the process of upgrading.

I think that's the same resolution given by @alex-thewsey-ibm. Deleting that particular dependency from yarn.lock  helped me.
Anyways, Thanks for this hack☺️

**Comment by felipecsl** [comment]
Using `resolutions` in `package.json` worked for me https://github.com/webpack/webpack-dev-server/issues/2739#issuecomment-695164486

**Comment by vfonic** [comment]
This should at least return some warning:

```sh
yarn add is-alphanumerical@1.0.0
yarn upgrade is-alphabetical
```

This is what I get instead (there are zero changes after this):

```
success Saved lockfile.
success Saved 0 new dependencies.
```

**Comment by mpdude** [comment]
For the record, GitHub's Dependabot will propose updates also for _transitive_ (indirect) dependencies if those contain known security issues.

https://dependabot.com/blog/securing-javascript-transitive-dependencies/

**Comment by olivierlacan** [comment]
@mpdude This is precisely why selective transitive dependency upgrading not being possible in Yarn (or NPM) is so problematic. We now have the largest code hosting service in the world (GitHub) recommending transitive dependency security updates that can't be achieved without complicated and [error-prone manual manipulations of the lockfile](https://github.com/yarnpkg/yarn/issues/4986#issuecomment-395036563) or [git history trickery](https://github.com/webpack/webpack-dev-server/issues/2739#issuecomment-695164486). 

This problem is only going to grow exponentially as more people get alerts from their repos and security scanning software (within companies) and business goals start prioritizing staying on top of attack vectors more and more. 

There is a pattern for how to do this successfully: [`bundle update <dependency>`](https://bundler.io/man/bundle-update.1.html) will try to resolve a new dependency graph regardless of whether `<dependency>` is a direct or indirect dependency and regardless of the version resolved for it in the lockfile. Bundler even offers a `--conservative` flag to avoid indirect dependency updates.

I'm curious why yarn can't be made to follow the same pattern. I'm not trying to diminish the work involved, I'm sure it's far beyond my understanding since I'm about as inexperienced in graph traversal as anyone can be, but it's concerning to not see a strong focus on hashing out a plan to get this feature tackled and eventually added. Nearly half

**Comment by shamilovtim** [comment]
Is pinning all versions to their exact versions, doing `yarn install` and then deleting the lock file immediately afterwards not a good solution?

**Comment by moroine** [comment]
I don't know why https://github.com/djmitche/yarn-minify feature is not built-in `yarn update` command 🤔

**Comment by djmitche** [comment]
..because it's a hack :)

Implementing that within yarn itself, in an appropriate fashion, is much harder!

**Comment by ExplodingCabbage** [comment]
2026 edit: I belatedly realise my answer below relates to Berry, not Classic. I had not appreciated in 2022 that the two versions had separate GitHub repos, and that this issue is on the repo for the unmaintained "Classic" version. I leave my comment below regardless.

---

The `--recursive` flag, as in `yarn up whatever-package --recursive`, appears to make `yarn up` happy to upgrade `whatever-package` even if it's a sub-dependency. I'm not sure if this is a bug or not, since my understanding of https://yarnpkg.com/cli/up#options-R%2C-recursive is that the main intent of `--recursive` is to let you upgrade a specified package's dependencies, not to let you upgrade a specified indirect dependency. But regardless I've found it was good enough for my purposes everywhere I've needed to do an indrect dependency upgrade so far!

**Comment by Tofandel** [comment]
From the linked issue, it seems this has been resolved already and running `yarn upgrade` should update indirect dependencies

**Comment by jrochkind** [comment]
Hi @Tofandel , can you say which linked issue you are referring to? There are several here, I think. Do you know what version of yarn resolves the issue?

Have you tried it out yourself to verify, and if so what version of yarn?

It's not working for me, but I have an old version of yarn. I am interested in what version of yarn might resolve it.

**Comment by Tofandel** [comment]
I'm referring to https://github.com/yarnpkg/yarn/issues/2394#issuecomment-386290827

I don't know which version, but yes for me doing `yarn upgrade` produces the same as deleting an old `yarn.lock` and running `yarn install`

**Comment by niklasholm** [comment]
@Tofandel `yarn upgrade` upgrades all dependencies, yes. This issue is about upgrading specific transitive dependencies without upgrading everything.

**Comment by Tofandel** [comment]
@niklasholm I'm afraid you need to reread the original issue then, nowhere does it says "without upgrading everything"

You can give a try with this package.json and this lock file
```json
{
  "dependencies": {
    "is-alphanumerical": "1.0.0"
  }
}
```
```
# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1

is-alphabetical@^1.0.0:
  version "1.0.0"
  resolved "https://registry.yarnpkg.com/is-alphabetical/-/is-alphabetical-1.0.0.tgz#e2544c13058255f2144cb757066cd3342a1c8c46"
  integrity sha512-qCTN0ODfqBekrQXyHiilXfcZA7hWs3wswkg5jmlKkTVI8IcFwIn33d0kBHpkZGp2/cYNVfASeP/mt/4Dgs84FQ==

is-alphanumerical@1.0.0:
  version "1.0.0"
  resolved "https://registry.yarnpkg.com/is-alphanumerical/-/is-alphanumerical-1.0.0.tgz#e06492e719c1bf15dec239e4f1af5f67b4d6e7bf"
  integrity sha512-vBnURR3j92CuvpwwAHs3nmttKsLPb04UyiSkeJquAaiMXj18Y+T6nBIFOSbBIbbsu5DViJeHdeAy5Sk7Se5S/Q==
  dependencies:
    is-alphabetical "^1.0.0"
    is-decimal "^1.0.0"

is-decimal@^1.0.0:
  version "1.0.4"
  resolved "https://registry.yarnpkg.com/is-decimal/-/is-decimal-1.0.4.tgz#65a3a5958a1c5b63a706e1b333d7cd9f630d3fa5"
  integrity sha512-RGdriMmQQvZ2aqaQq3awNA6dCGtKpiDFcOzrTWrDAT2MiWrKQVPmxLGHl7Y2nNu6led0kEyoX0enY0qXYsv9zw==
```

Run yarn upgrade

```
# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


is-alphabetical@^1.0.0:
  version "1.0.4"
  resolved "https://registry.yarnpkg.com/is-alphabetical/-/is-alphabetical-1.0.4.tgz#9e7d6b94916be22153745d184c298cbf986a686d"
  integrity sha512-DwzsA04LQ10FHTZuL0/grVDk4rFoVH1pjAToYwBrHSxcrBIGQuXrQMtD5U1b0U2XVgKZCTLLP8u2Qxqhy3l2Vg==

is-alphanumerical@1.0.0:
  version "1.0.0"
  resolved "https://registry.yarnpkg.com/is-alphanumerical/-/is-alphanumerical-1.0.0.tgz#e06492e719c1bf15dec239e4f1af5f67b4d6e7bf"
  integrity sha512-vBnURR3j92CuvpwwAHs3nmttKsLPb04UyiSkeJquAaiMXj18Y+T6nBIFOSbBIbbsu5DViJeHdeAy5Sk7Se5S/Q==
  dependencies:
    is-alphabetical "^1.0.0"
    is-decimal "^1.0.0"

is-decimal@^1.0.0:
  version "1.0.4"
  resolved "https://registry.yarnpkg.com/is-decimal/-/is-decimal-1.0.4.tgz#65a3a5958a1c5b63a706e1b333d7cd9f630d3fa5"
  integrity sha512-RGdriMmQQvZ2aqaQq3awNA6dCGtKpiDFcOzrTWrDAT2MiWrKQVPmxLGHl7Y2nNu6led0kEyoX0enY0qXYsv9zw==
```

As you can see is-alphabetical was bumped to version 1.0.4

CQFD this issue can be closed

**Comment by jrochkind** [comment]
The entire discussion in this ticket the entire time has assumed "without upgrading everything".  If you wanted to upgrade everything, indeed `yarn upgrade` (without arguments) would have done it, even back in 2017 when this ticket was filed. I think everyone knew that, there was no reason for the ticket to be filed unless it meant "without upgrading everything" so it was assumed without being said explicit in 2017. 

Meanwhile, I still am confused about whether this now works in yarn, and if so introduced in what version. 

but I don't think I'm ever going to learn it from this ticket, I'll try to find some time to experiment.

**Comment by Tofandel** [comment]
@jrochkind It still doesn't then, I just tried it

**Comment by ndp** [comment]
FWIW, I was able to upgrade an indirect dependency using the `--recursive` flag. In my case, I want to upgrade `loader-utils`, which is not in my `package. json` but has a security alert. I was able to run:

    yarn up --recursive loader-utils

It modified my `yarn.lock`, but nothing else. In fact, I was able to take care of three security vulnerabilities this way. For me, this is a workable solution and could simply be documented as the solution unless there are cases that don't work.

   $ yarn --version
   3.2.0

**Comment by sbernard31** [comment]
If you are using old stable version (e.g. 1.22.x) : I didn't find better solution than : https://medium.com/@ayushya/upgrading-javascript-packages-deep-dependencies-using-yarn-8b5983d5fb6b

**Comment by matthewtusker** [comment]
> If you are using old stable version (e.g. 1.22.x) : I didn't find better solution than : https://medium.com/@ayushya/upgrading-javascript-packages-deep-dependencies-using-yarn-8b5983d5fb6b

This works far better than any of the solutions listed here. `yarn up --recursive loader-utils` upgrades `loader-utils`, but has also upgraded a bunch of other stuff, and has still left to disparate versions. Deleting the lines from the lock file and then running `yarn install` installs a single version.

**Comment by pftg** [comment]
> > If you are using old stable version (e.g. 1.22.x) : I didn't find better solution than : https://medium.com/@ayushya/upgrading-javascript-packages-deep-dependencies-using-yarn-8b5983d5fb6b
> 
> This works far better than any of the solutions listed here. `yarn up --recursive loader-utils` upgrades `loader-utils`, but has also upgraded a bunch of other stuff, and has still left to disparate versions. Deleting the lines from the lock file and then running `yarn install` installs a single version.

Here is script which I used to do the same: https://github.com/yarnpkg/yarn/issues/4986#issuecomment-562719589

**Comment by microJ** [comment]
> If you are using old stable version (e.g. 1.22.x) : I didn't find better solution than : https://medium.com/@ayushya/upgrading-javascript-packages-deep-dependencies-using-yarn-8b5983d5fb6b

work for me.
simply, I want to upgrade react-native@0.63.5, then run `yarn remove react-native && yarn add react-native@0.63.5` work for me well

**Comment by gonadarian** [comment]
I use the following "clean" approach (without touching `yarn.lock`) in such situations.

If, as described in original example, we have:

- `yarn add is-alphanumerical@1.0.0`
    - installs `is-alphabetical` **v1.0.1**

I would do these 3 steps:

1. `yarn add is-alphabetical@1.0.2`
    - gives **duplicated** `is-alphabetical` with both original **v1.0.1** & new **v1.0.2** in `yarn.lock`
2. `npx yarn-deduplicate`
    - deduplicates to **only** the latest `is-alphabetical` **v1.0.2**
3. yarn remove is-alphabetical
    - cleans up `package.json`
4. profit!

**Comment by romainmenke** [comment]
We switched the entire release strategy of `postcss-preset-env` to accommodate `yarn` but in hindsight I think this was a mistake. We should have left the issue as is as it wasn't ours to fix.

Our current setup :
- build a dependency graph of all parts of `postcss-preset-env`
- check if any part needs to be released
- release only those parts that don't themselves depend on unreleased parts
- update all downstream dependents with the newly released part's version
- go to step 1

We repeat this until the entire graph has been updated and released.

Obviously a small change in a deeply buried part (e.g. a tokenizer) escalates to ±50 releases of packages that didn't themselves have any code changes.

This in turn creates a lot of noice for our users.

We don't like this solution and ideally we wouldn't have to do stuff like this.

We are only doing this because `yarn` fails to correctly update transitive dependencies, leading to conflicts when peer dependencies are used.

No other package manager is affected by this issue.

**Comment by ClementValot** [comment]
@rjimenezda Quick idea there for a workaround, you may be able to leverage `yarn dedupe`

> yarn add vite@latest
> yarn dedupe 
> yarn remove vite

Would install the latest vite version, then deduplicate the vite dependency along semver to only keep the latest, then remove the explicit dependency. 

Also, have you checked whether somewhere in your project you would have another transitive dependency to vite which prevents upgrading to somthing higher than 5.0.8? `yarn why vite` should help in that case

**Comment by robertknight** [comment]
I wrote a dumb tool to automate the process of editing `yarn.lock` to remove the existing resolutions and then re-running `yarn install`: https://github.com/robertknight/yarn-update-indirect.

```
npm install yarn-update-indirect
yarn-update-indirect some-transitive-dep
```

**Comment by dobesv** [comment]
BTW in yarn berry what seems to work is looking up the latest version and running `yarn set resolution`, e.g.:

    yarn set resolution "node-gyp@npm:latest" "npm:11.2.0"

**Comment by n0099** [comment]
If you are using `yarn^3`, try `yarn up -R`: https://github.com/yarnpkg/berry/issues/2042
